redhat-appstudio/build-service
# Allow multiple credentials secrets in single namespace

## Summary

This PR allows the usage of mutiple build secrets per SCM type, host or repository.

It supports the `ssh-auth` and `basic-auth` type secrets.

Secrets must have the `appstudio.redhat.com/credentials: scm` and `appstudio.redhat.com/scm.host` labels, as well as might have an optional `appstudio.redhat.com/scm.repository` annotation value.

The best suitable one is selected, if none match then fallback is GitHub APP secret in the global namespace.

Known limitations:
 - GitHub dropped the login + password auth, seems only tokens are possible for now;
 - SSH key login isn't supported by any client, will need to find the way to perform authentication using it.

## Issue to Fix

https://issues.redhat.com/browse/SVPI-745
  
## How to test


### Prerequisites:
 - GitHub organization
 - PAT with repo, user, org permissions;
 - Quay organization and repository with OAuth app enabled, known OAuth token for it.
  - Build-service fork and custom image. (typically, `mshaposh/build-controller:latest` is what in th PR)
 - Openshift cluster
 
### Installation
 - Get the latest version of `infra-deployments`
 - Create `hack/preview.env` from template. 
   Set the necessary fields:
   
   ```
   // local name of the fork fremote
   export MY_GIT_FORK_REMOTE=mshaposh
   ...
   // GH org and token
   export MY_GITHUB_ORG=mshaposhnik-org
   export MY_GITHUB_TOKEN=ghp_psj......
   ...
   // Quay org and token
   export IMAGE_CONTROLLER_QUAY_ORG=uy7c
   export IMAGE_CONTROLLER_QUAY_TOKEN=J8cM177j6......
   ...
   // Build service image and tag
   export BUILD_SERVICE_IMAGE_REPO=mshaposh/build-controller
   export BUILD_SERVICE_IMAGE_TAG=latest
   ```

 - Make sure `oc` is logged into a cluster, run from the infra-deployment project root:
  `./hack/bootstrap-cluster.sh preview --keycloak --toolchain`
  Make sure no fundamental errors during installlation.
  
 ### Testing
  - Create application from minimal template:
  ```
oc create -f - << EOF 
apiVersion: appstudio.redhat.com/v1alpha1
kind: Application
metadata:
  annotations:
  name: go-application-sample
  namespace: default
spec:
  appModelRepository:
    url: ''
  displayName: go-application-sample
  gitOpsRepository:
    url: ''
EOF
```
Wait it for resolve (must have status similar to this:)
```
 conditions:
    - lastTransitionTime: '2024-02-15T11:56:37Z'
      message: Application has been successfully created
      reason: OK
      status: 'True'
      type: Created
  devfile: |
    metadata:
      attributes:
        appModelRepository.context: /
        appModelRepository.url: https://github.com/mshaposhnik-org/go-application-sample-N6juw-marry-sign
        gitOpsRepository.context: ./
        gitOpsRepository.url: https://github.com/mshaposhnik-org/go-application-sample-N6juw-marry-sign
      name: go-application-sample
    schemaVersion: 2.2.0
```


- Create the SCM secret  (set or omit the repository annotation) :
```

apiVersion: v1
kind: Secret
metadata:
  namespace: default
  name: test-scm-secret
  labels:
    appstudio.redhat.com/credentials: scm
    appstudio.redhat.com/scm.host: github.com
  annotations:
    appstudio.redhat.com/scm.repository: mshaposhnik/*
type: kubernetes.io/basic-auth
stringData:
  password: Z2hw....
EOF
```

- Create component:
```
oc create -f - << EOF 
apiVersion: appstudio.redhat.com/v1alpha1
kind: Component
metadata:
  name: go-component-sample
  namespace: default
  annotations:
    build.appstudio.openshift.io/request: "configure-pac"
    image.redhat.com/generate: '{"visibility": "public"}'
spec:
  componentName: go-sample-component-name
  application: go-application-sample
  source:
    git:
      url: https://github.com/mshaposhnik/devfile-sample-go-basic.git
EOF
      
```
It must resolve status:
```
status:
  conditions:
    - lastTransitionTime: '2024-02-15T12:36:44Z'
      message: GitOps resource generated successfully
      reason: OK
      status: 'True'
      type: GitOpsResourcesGenerated
    - lastTransitionTime: '2024-02-15T12:36:44Z'
      message: Component has been successfully created
      reason: OK
      status: 'True'
      type: Created
```

PR to repository must be submitted.
